service-worker.js

```diff
@@ -23,117 +23,191 @@
+// See equivalent regex in app.js function refreshCacheStatus() and ensure both are the same
+var excludedURLSchema = /^(?:chrome-extension|example-extension):/i;
+
+// Pattern for ZIM file namespace - see https://wiki.openzim.org/wiki/ZIM_file_format#Namespaces
+// In our case, there is also the ZIM file name, used as a prefix in the URL
+var regexpZIMUrlWithNamespace = /(?:^|\/)([^\/]+\/)([-ABIJMUVWX])\/(.+)/;
+
+self.addEventListener('install', function (event) {
     event.waitUntil(self.skipWaiting());
 });
 
-self.addEventListener('activate', function(event) {
+self.addEventListener('activate', function (event) {
     // "Claiming" the ServiceWorker is necessary to make it work right away,
     // without the need to reload the page.
     // See https://developer.mozilla.org/en-US/docs/Web/API/Clients/claim
     event.waitUntil(self.clients.claim());
 });
 
-var regexpRemoveUrlParameters = new RegExp(/([^?#]+)[?#].*$/);
+var outgoingMessagePort = null;
+var fetchCaptureEnabled = false;
 
-// This function is duplicated from uiUtil.js
-// because using requirejs would force to add the 'fetch' event listener
-// after the initial evaluation of this script, which is not supported any more
-// in recent versions of the browsers.
-// Cf https://bugzilla.mozilla.org/show_bug.cgi?id=1181127
-// TODO : find a way to avoid this duplication
+self.addEventListener('fetch', function (event) {
+    if (fetchCaptureEnabled &&
+        regexpZIMUrlWithNamespace.test(event.request.url) &&
+        event.request.method === "GET") {
 
-/**
- * Removes parameters and anchors from a URL
- * @param {type} url
- * @returns {String} same URL without its parameters and anchors
- */
-function removeUrlParameters(url) {
-    return url.replace(regexpRemoveUrlParameters, "$1");
-}
+        // The ServiceWorker will handle this request either from CACHE or from app.js
 
-var outgoingMessagePort = null;
-var fetchCaptureEnabled = false;
-self.addEventListener('fetch', fetchEventListener);
+        event.respondWith(
+            // First see if the content is in the cache
+            fromCache(event.request).then(
+                function (response) {
+                    // The response was found in the cache so we respond with it 
+                    return response;
+                },
+                function () {
+                    // The response was not found in the cache so we look for it in the ZIM
+                    // and add it to the cache if it is an asset type (css or js)
+                    return fetchRequestFromZIM(event).then(function (response) {
+                        // Add css or js assets to CACHE (or update their cache entries) unless the URL schema is not supported
+                        if (cachedContentTypesRegexp.test(response.headers.get('Content-Type')) &&
+                            !excludedURLSchema.test(event.request.url)) {
+                            event.waitUntil(updateCache(event.request, response.clone()));
```

Comment:
Why is it necessary to wait for the cache to be updated before returning the response? (2 lines below).
It should not be a problem that the cache can be updated a bit later, and should speed up the request/response timing?

Reply:
@mossroy It doesn't wait. `event.waitUntil` may look like async `await`, but in fact it is merely an instruction to the browser not to let the event expire until the cache has been updated (events are expired quickly by browsers to optimize memory usage). The next line (`return response;`) will run immediately, before the `updateCache` promise completes, and we can't `return` before the `event.waitUntil` statement or else that code will become unreachable.

Reply:
All right, thanks for the explanation